### PR TITLE
[WIP] add new APIs to build dataset

### PR DESCRIPTION
**Pull request by zhangguanheng66**
With some very basic APIs (download_from_url, extract_archive, build_dictionary, rationed_split, generate_ngrams), we hope to build new datasets directly from torch.utils.data.dataset. The existing abstract dataset class in TorchText (i.e. torchtext.data.Dataset) is too complicated. We hope to gain flexibility with this simplified pattern. A few supervised learning dataset will be implemented based on this new pattern. As more necessary APIs are added, we hope the new APIs allow a user to do everything they want to do.

Steps to build datasets with the new pattern:
- The download_from_url and extract_archive functions are used to download and extract the online raw data. 
- Some text preprocessors/processors (e.g. spaCy) are required to clean the data and tokenize the sentence. It should be noted that the text preprocessors vary with the actual datasets so users are expected to prepare the preprocessors/processors. A generic preprocessor API is not pursued here.
- The build_vocab function is to build vocab object based on the tokens lists.
- Use rationed_split function to split the initial data into train/test/valid sets.
- The train/test/valid sets generated from last step are passed to the generate_iterators function. Here, data are batched. The output (train/test/valid iterators) can be used to train/test models.

### CODE REVIEW

File: torchtext/data/iterator.py [hunk 1]
```diff
@@ -252,6 +252,29 @@ def create_batches(self):
                                 sort_within_batch=self.sort_within_batch)
 
 
+def generate_iterators(datasets, batch_sizes, **kwargs):
+    """Create Iterator objects for multiple splits of a dataset.
+
+    Arguments:
+        datasets: a tuple of Datasets
+        batch_sizes: a list of batch_size, corresponding to each dataset.
+        ngrams: the number of ngrams.
```
Comment (cpuhrsch) [inline torchtext/data/iterator.py]: This is specific to text, but the function has potential for being generic by itself. Is there a way to split out this "ngrams" concept?

Comment (zhangguanheng66) [reply torchtext/data/iterator.py]: It seems a typo. I don't think iterator function needs ngrams.

File: torchtext/utils.py [hunk 2]
```diff
@@ -25,7 +27,18 @@ def inner(b=1, bsize=1, tsize=None):
 
 
 def download_from_url(url, path):
-    """Download file, with logic (from tensor2tensor) for Google Drive"""
+    """Download file, with logic (from tensor2tensor) for Google Drive
```
Comment (cpuhrsch) [inline torchtext/utils.py]: We could take this as a chance to move this stuff into pytorch core and then use it here.

Comment (zhangguanheng66) [reply torchtext/utils.py]: yeah. As long as they are not text specific (I think most of them are not), we can move them to the core library.

File: text_classification.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+        self.fields = fields
+
+    def __getitem__(self, i):
+        return self.examples[i]
+
+    def __len__(self):
+        try:
+            return len(self.examples)
+        except TypeError:
+            return 2**32
```
Comment (cpuhrsch) [inline text_classification.py]: Does this mean self.examples could be something else than a list? Also, why call it examples, isn't it the entire corpus?

File: text_classification.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+                'float' token will be used.
+            ngrams: a contiguous sequence of n items from s string text.
+                Default: 1
+
+        Examples:
+            >>> txt_cls = TextClassificationDataset("ag_news", ngrams=2)
+
+        """
+        self.dataset_name = dataset_name
+        if dataset_name in supported_text_classification_datasets.keys():
```
Comment (cpuhrsch) [inline text_classification.py]: You have the class instances below (e.g. AG_NEWS), so you use this to deduplicate some code I think.
 The main reason for that seems to be that you're using "load_text_classification_data" and have a predefined split.

File: text_classification.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+            - valid_iter: a iterator based on
+                train.csv file with 1-split_ratio percent.
+
+        """
+
+        train, test, valid = self.split(split_ratio, random_state)
+        return generate_iterators(
+            (train, test, valid), [batch_size] * 3, device=device)
+
+    def download(self):
```
Comment (cpuhrsch) [inline text_classification.py]: Can we move this out of TextClassificationDataset?

File: text_classification.py [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+        os.makedirs(self.raw_folder)
+        filename = self.dataset_name + '_csv.tar.gz'
+        url = self.url
+        path = os.path.join(self.raw_folder, filename)
+        download_from_url(url, path)
+        extract_archive(path, self.raw_folder, remove_finished=True)
+
+        print('Dataset %s downloaded.' % self.dataset_name)
+
+    def preprocess(self, raw_folder=None, processed_folder=None):
```
Comment (cpuhrsch) [inline text_classification.py]: This seems useful in its own right, i.e. it applies a preprocessing function to a set of text files. Can we move it out?

File: text_classification.py [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+
+        src_filepath = os.path.join(raw_folder, 'test.csv')
+        if not os.path.isfile(src_filepath):
+            self.download()
+        tgt_filepath = os.path.join(processed_folder, self.dataset_name + '.test')
+        self.text_normalize(src_filepath, tgt_filepath)
+
+        print("Dataset %s preprocessed." % self.dataset_name)
+
+    def text_normalize(self, src_filepath, tgt_filepath):
```
Comment (cpuhrsch) [inline text_classification.py]: Can we move this out of the class and make it a general function? This seems useful by itself. We just need to add a bit of an explanation I think (i.e. what kind of text this is useful for).

File: text_classification.py [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, dataset_name, root='.data',
+                 text_field=None, label_field=None, ngrams=1):
```
Comment (cpuhrsch) [inline text_classification.py]: The number of ngrams is passed in as a parameter for load_text_classification_data, which in turn calls into label_text_processor and returns the preprocessed "examples" (corpus). build_vocab then takes an instance of self.  I agree that using the iterator approach is more efficient, but in this instance it is iterating over a corpus that's already entirely in memory. So, when build the vocabulary it should be enough to only pass in self.examples. Then build_vocab could stand entirely by itself (maybe even called "build_vocab_from_list") .

File: text_classification.py [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+                                             'processed')
+
+        filepath = os.path.join(self.processed_folder, self.dataset_name + '.train')
+        self.train_examples = self.load_text_classification_data(filepath, ngrams)
+
+        filepath = os.path.join(self.processed_folder, self.dataset_name + '.test')
+        self.test_examples = self.load_text_classification_data(filepath, ngrams)
+        self.examples = self.train_examples + self.test_examples
+
+        self.fields['text'].vocab = build_vocab(self, self.fields['text'], 'text')
```
Comment (cpuhrsch) [inline text_classification.py]: I think we can call into this using self.examples. Then build_vocab can be more generic . This however also requires us to get rid of "fields".

File: text_classification.py (hunk 3 above)
Comment (zhangguanheng66) [reply text_classification.py]: This is the data loaded from train/test files. It covers the entire corpus in this case.

File: text_classification.py (hunk 4 above)
Comment (zhangguanheng66) [reply text_classification.py]: This interface is old, and not generic. The major function components in this class should be "download", "preprocess", "load token list", "general iterator/batch". That's all.

File: text_classification.py (hunk 9 above)
Comment (zhangguanheng66) [reply text_classification.py]: I will try to improve the build_vocab API. For now, I think two separate components, data (e.g. examples) and token processor (e.g. field) make sense.

File: text_classification.py (hunk 5 above)
Comment (zhangguanheng66) [reply text_classification.py]: yeap, download function is now independent and can stay outside.

File: text_classification.py (hunk 6 above)
Comment (zhangguanheng66) [reply text_classification.py]: Similar to download, we don't need those wrappers inside the class.

File: text_classification.py (hunk 7 above)
Comment (zhangguanheng66) [reply text_classification.py]: Same as above

**Comment by zhangguanheng66** [comment]
Split the supervised learning dataset into a new PR. https://github.com/pytorch/text/pull/557
Land this one first for the necessary APIs.

File: torchtext/vocab.py [hunk 10]
```diff
@@ -216,6 +218,33 @@ def set_vectors(self, stoi, vectors, dim, unk_init=torch.Tensor.zero_):
...
+        if not field.sequential:
+            x = [x]
+        try:
+            counter.update(x)
+        except TypeError:
+            counter.update(chain.from_iterable(x.text))
+    specials = list(OrderedDict.fromkeys(
+        tok for tok in [field.unk_token, field.pad_token, field.init_token,
+                        field.eos_token] if tok is not None))
+    return Vocab(counter, specials=specials, **kwargs)
```
Comment (cpuhrsch) [inline torchtext/vocab.py]: Vocab does a lot of things that seem unrelated to dictionaries, for example "load_vectors" and "set_vectors", "__eq__" etc.  - maybe we should audit it and see if we can remove a few things for 0.4?

Vocab is also undocumented and not referenced by anything, even though it seems pretty useful.

Comment (zhangguanheng66) [reply torchtext/vocab.py]: Let me see if I can make some changes there. I cannot estimate the efforts to split vacab and vector etc.

File: torchtext/utils.py [hunk 11]
```diff
@@ -75,3 +88,31 @@ def unicode_csv_reader(unicode_csv_data, **kwargs):
 def utf_8_encoder(unicode_csv_data):
     for line in unicode_csv_data:
         yield line.encode('utf-8')
+
+
+def extract_archive(from_path, to_path=None, remove_finished=False):
```
Comment (cpuhrsch) [inline torchtext/utils.py]: Add a flag like "format=tar" to indicate it's using tar so we can extend it later on. This is also very useful in general.

File: torchtext/data/iterator.py [hunk 12]
```diff
@@ -252,6 +252,28 @@ def create_batches(self):
...
+    Examples:
+        >>> batch_size = 64
+        >>> train_iter, test_iter, valid_iter = generate_iterators(
+            (train_dataset, test_dataset, valid_dataset), [batch_size] * 3,
+            device="cpu")
+    """
+    ret = []
+    for i in range(len(datasets)):
+        train = i == 0
+        ret.append(Iterator(
```
Comment (cpuhrsch) [inline torchtext/data/iterator.py]: It would be nice if we didn't need to use any existing legacy stuff since we're planning to remove / revamp it

File: torchtext/data/iterator.py [hunk 13]
```diff
@@ -252,6 +252,28 @@ def create_batches(self):
                                 sort_within_batch=self.sort_within_batch)
 
 
+def generate_iterators(datasets, batch_sizes, **kwargs):
```
Comment (cpuhrsch) [inline torchtext/data/iterator.py]: Let's write model without batching first, because batching is a perf. optimization and we haven't measured it yet.

File: torchtext/vocab.py (hunk 10 above)
Comment (cpuhrsch) [reply torchtext/vocab.py]: It's ok if we build a very minimal and advertised Dictionary instead of replicating all features of Vocab. That will also give us the chance to make it fast.

File: torchtext/utils.py (hunk 11 above)
Comment (zhangguanheng66) [reply torchtext/utils.py]: We can simply extend it by adding the other branches between "if" and "else"

File: torchtext/data/iterator.py (hunk 12 above)
Comment (zhangguanheng66) [reply torchtext/data/iterator.py]: Then, I will move this to tutorial.

File: torchtext/vocab.py [hunk 14]
```diff
@@ -216,6 +218,34 @@ def set_vectors(self, stoi, vectors, dim, unk_init=torch.Tensor.zero_):
                 self.vectors[i] = unk_init(self.vectors[i])
 
 
+def build_dictionary(dataset, field, data_name, **kwargs):
+    """Construct the Vocab object for the field from a dataset.
```
Comment (cpuhrsch) [inline torchtext/vocab.py]: Still need to update the docs :)

File: torchtext/data/dataset.py [hunk 15]
```diff
@@ -323,8 +323,22 @@ def stratify(examples, strata_field):
 
 
 def rationed_split(examples, train_ratio, test_ratio, val_ratio, rnd):
-    # Create a random permutation of examples, then split them
-    # by ratio x length slices for each of the train/test/dev? splits
+    """Create a random permutation of examples, then split them by ratios
```
Comment (fmassa) [inline torchtext/data/dataset.py]: This looks a lot like [`torch.utils.data.random_split`](https://github.com/pytorch/pytorch/blob/f622ec8084c1778c6943474796efa52e9b9cf695/torch/utils/data/dataset.py#L262), but where the splits don't need necessarily to be random.

File: torchtext/data/dataset.py [hunk 16]
```diff
@@ -323,8 +323,22 @@ def stratify(examples, strata_field):
...
+
+    Arguments:
+        examples: a list of data
+        train_ratio, test_ratio, val_ratio: split fractions.
+        rnd: a random shuffler
+
+    Examples:
+        >>> examples = []
+        >>> train_ratio, test_ratio, val_ratio = 0.7, 0.2, 0.1
+        >>> rnd = torchtext.data.dataset.RandomShuffler(None)
```
Comment (fmassa) [inline torchtext/data/dataset.py]: Why do we re-create a `RandomShuffler` in `torchtext`, what is the use-case, instead of using `torch.utils.data.RandomShuffler`?
